### ISSUE TITLE
[MKS] Support Public Cloud Loadbalancer as load balancing solution

### ISSUE THREAD

**Issue by antonin-a** [ovh/public-cloud-roadmap#603]
As a MKS user I would like to be able to use [Public Cloud Load Balancer](https://www.ovhcloud.com/en/public-cloud/load-balancer/) as load balancing solution for my MKS clusters. 
So that I can benefit from the performances and features of Public Cloud Loadbalancer.

**Comment by antonin-a** [comment]
Our historical load balancer [Loadbalancer for Managed Kubernetes](https://www.ovhcloud.com/en/public-cloud/load-balancer-kubernetes/) was dedicated to our Managed Kubernetes Service, at the end of 2023 we launched the [Public Cloud Load Balancer](https://www.ovhcloud.com/en/public-cloud/load-balancer/) that was usable to load balance traffic on your public cloud instances, it is now also supported for your Managed Kubernetes Service nodes.  

This means you can use a single solution to manage load balancing on both your public cloud instances and your Managed Kubernetes Service clusters.  

Based on the Octavia project, Public Cloud Loadbalancer is a high-performance, open-source load balancing solution part of the  OpenStack Foundation. It is designed to provide advanced load balancing features, enabling optimized traffic management to your applications hosted on our Public Cloud. 

Here is a short comparison table: 
|                                | Load balancer for Kubernetes | Public Cloud Load Balancer|
|-----|----|---|
|Max number of connections               | 10 000                     | up to 40 000*                 |
| Max number of HTTP requests          | 2 000                      | up to 10 000*                  |
| Bandwidth                                           | 200 Mbit/s              | up to 2 Gbit/s(up/down) |
| Supported protocol                            | TCP                         | TCP/UDP                         |
| Supported load balancing layers       | L4                            | L4 / L7**                         |
| Capacity to export metrics and logs | ❌                            | ✔️                                    |
| Private to private scenario                | ❌                            | ✔️                                    |
| Floating IP                                          | ❌                            | ✔️                                    |

*current limitation with L plan, larger plans to come.
**using  Openstack Octavia ingress-controller component. 

### **Important announcement:**
For MKS cluster running kubernetes version 1.31 and above, the [Public Cloud Load Balancer](https://www.ovhcloud.com/en/public-cloud/load-balancer/) will be the default load balancing solution.  

The is no impact on existing services and clusters. For clusters running Kubernetes versions 1.31 and above, the deployment/redeployment of a Kubernetes manifest for a 'Loadbalancer' service without specifying a 'class', will create a Public Cloud Loadbalancer , for more details you can refer to [our documentation](https://help.ovhcloud.com/csm/en-gb-public-cloud-kubernetes-expose-applications-using-load-balancer?id=kb_article_view&sysparm_article=KB0062867).


### **Breaking change:**
Unlike the [Loadbalancer for Managed Kubernetes](https://www.ovhcloud.com/en/public-cloud/load-balancer-kubernetes/), with the [Public Cloud Load Balancer](https://www.ovhcloud.com/en/public-cloud/load-balancer/), if you wish to expose your LoadBalancer publicly, it is mandatory use a subnet that has an OVHcloud Gateway (or an IP address available for creating one). For more information, and to make sure that your networks are correctly setup, please refer to the following [documentation](https://help.ovhcloud.com/csm/en-public-cloud-kubernetes-expose-applications-using-load-balancer?id=kb_article_view&sysparm_article=KB0062873#network-prerequisite-to-expose-your-load-balancers-publicly).

**Comment by antonin-a** [comment]
Good point @matmicro, no impact for existing users part of the Beta phase that are already using Public Cloud Load Balancer. You can keep the 'octavia' annotation or remove from your manifest on 1.31, the behavior will be the same.

**Comment by rverchere** [comment]
Hi,

I'm starting migrating to this LB, and have some questions / remarks. I'll post here to have some follow-up.

I'm using ingress-nginx controller, which creates the LB resource, and I have just updated the helm values to manage the [replacement](https://help.ovhcloud.com/csm/en-ie-public-cloud-kubernetes-migrate-iolb-to-public-cloud-loadbalancer?id=kb_article_view&sysparm_article=KB0066711#replacement). The new LB has been created with all the related resources.

For now, my cluster is full public (no vrack attached).

## Getting the source IP behind the LoadBalancer

Following [this article](https://help.ovhcloud.com/csm/fr-public-cloud-kubernetes-getting-source-ip-behind-loadbalancer?id=kb_article_view&sysparm_article=KB0055065), I have updated the annotation to use `loadbalancer.openstack.org/proxy-protocol`.

The old metadata `proxy-real-ip-cidr` based on `lb.k8s.ovh.net/egress-ips` seems not accurate anymore. As I am in a full public MKS, I had to change the `proxy-real-ip-cidr` to the public IP of the gateway created by the LoadBalancer.

So:
- Is  `lb.k8s.ovh.net/egress-ips` still used by Octavia LB ?
- Can you conform / update the article on getting sources to manage different MKS network configuration (public, attached to a vrack)?

## Getting Prometheus Metrics

When creating a Load Balancer from kubernetes, the listeners are automatically created / managed. I've seen this PR https://github.com/kubernetes/cloud-provider-openstack/pull/2633 to manage metrics from K8S.

What is the current status?

## Automatic vrack/gateway creation

When setting Octavia LB, a new vrack and gateway are created, as my cluster is public. Is there a way to use existing vrack/gateway? Or can we define it (names, vlan id, subnet, etc.), i.e. with Terraform resources?

**Comment by rverchere** [comment]
Replying to myself:

- Getting the source IP behind the LoadBalancer -> an annotation `loadbalancer.ovhcloud.com/egress-ips` is set to the resource
